modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/endpoint/AbstractAssetManagerRestEndpoint.java

```diff
@@ -278,13 +285,31 @@ public Response getMediaPackage(@PathParam("mediaPackageID") final String mediaP
               responseCode = SC_INTERNAL_SERVER_ERROR)})
   public Response getAsset(@PathParam("mediaPackageID") final String mediaPackageID,
                            @PathParam("mediaPackageElementID") final String mediaPackageElementID,
-                           @PathParam("version") final String version) {
+                           @PathParam("version") final String version,
+                           @HeaderParam("If-None-Match") String ifNoneMatch) {
+
     try {
       for (final Version v : getAssetManager().toVersion(version)) {
         for (Asset asset : getAssetManager().getAsset(v, mediaPackageID, mediaPackageElementID)) {
+
+          if (StringUtils.isNotBlank(ifNoneMatch)) {
+            Checksum checksum = asset.getChecksum();
+
+            if (checksum == null || !checksum.getType().equals((ChecksumType.DEFAULT_TYPE))) {
+              checksum = Checksum.create(ChecksumType.DEFAULT_TYPE, asset.getInputStream());
```

Comment:
@KatrinIhler Just thought about this again: 

I would recommend to not calculate the checksum in case it is missing but rather log a warning message ("checksum missing").
At the end of the day, not having a checksum is a data consistency error and just masking this error by recalculating it on-the-fly should not be necessary. Note that of more than 3 million assets in our production database, none has no checksum.

Reply:
I agree, I'm gonna change this.